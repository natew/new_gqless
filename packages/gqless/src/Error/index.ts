import type { GraphQLError } from 'graphql';

export class gqlessError extends Error {
  graphQLErrors?: ReadonlyArray<GraphQLError>;
  otherError?: unknown;

  constructor(
    message: string,
    {
      graphQLErrors,
      otherError,
      caller,
    }: {
      graphQLErrors?: gqlessError['graphQLErrors'];
      otherError?: gqlessError['otherError'];
      caller?: Function;
    } = {}
  ) {
    super(message);

    if (graphQLErrors) this.graphQLErrors = graphQLErrors;
    if (otherError !== undefined) this.otherError = otherError;

    /* istanbul ignore else */
    if (caller && Error.captureStackTrace!)
      Error.captureStackTrace(this, caller);
  }

  toJSON() {
    return {
      message: this.message,
      graphQLErrors: this.graphQLErrors,
      otherError: this.otherError,
    };
  }

  static create(error: unknown, caller?: Function): gqlessError {
    let newError: gqlessError;

    if (error instanceof gqlessError) newError = error;
    else if (error instanceof Error)
      newError = Object.assign(new gqlessError(error.message), error);
    else
      newError = new gqlessError('Unexpected error type', {
        otherError: error,
      });

    /* istanbul ignore else */
    if (caller && Error.captureStackTrace!)
      Error.captureStackTrace(newError, caller);

    return newError;
  }
}

export * from './retry';
