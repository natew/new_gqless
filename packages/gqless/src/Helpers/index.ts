export { selectFields } from './selectFields';
export type { HydrateCacheOptions } from './ssr';
