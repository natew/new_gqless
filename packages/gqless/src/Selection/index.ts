export * from './selection';
export type { SelectionManager } from './SelectionManager';
export type {
  BuildSelectionInput,
  BuildSelectionValue,
} from './SelectionBuilder';
