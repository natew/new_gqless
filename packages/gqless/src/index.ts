export * from './Client';
export * from './Selection';
export * from './Schema';
export * from './Error';
export * from './Poller';
export * from './Helpers';
export * from './Utils/cycle';
