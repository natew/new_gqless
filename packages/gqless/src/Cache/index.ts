import lodashGet from 'lodash/get';
import lodashMerge from 'lodash/merge';

import { Selection } from '../Selection';

export const CacheNotFound = Symbol('Not Found');

export function createCache() {
  const cache: Record<string, unknown> = {};

  function getCacheFromSelection(selection: Selection) {
    const path = selection.path.slice(1);
    return lodashGet(cache, path, CacheNotFound);
  }

  function mergeCache(data: unknown) {
    lodashMerge(cache, data);
  }

  return { cache, getCacheFromSelection, mergeCache };
}
