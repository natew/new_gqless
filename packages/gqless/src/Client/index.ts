import { CacheNotFound, createCache } from "../Cache";
import { buildQuery } from "../QueryBuilder";
import { Selection } from "../Selection/selection";
import { QueryFetcher, ScalarsHash, Schema } from "../types";
import { deferredPromise } from "../Utils/promise";

const ProxySymbol = Symbol("gqless-proxy");

export function createClient<GeneratedSchema = never>(
  schema: Readonly<Schema>,
  scalars: ScalarsHash,
  queryFetcher: QueryFetcher
) {
  const ProxyCache = new WeakMap<Selection, object>();
  const ProxyCacheReverse = new WeakMap<object, Selection>();

  const ProxyResolvePromises = new WeakMap<object, Promise<void>>();

  const globalSelections = new Set<Selection>();
  const client: GeneratedSchema = createSchemaProxy();
  const { getCacheFromSelection, mergeCache } = createCache();

  function resolve(proxy: object): Promise<void> {
    return (
      ProxyResolvePromises.get(proxy) ||
      (() => {
        const proxySelection = ProxyCacheReverse.get(proxy);
        if (!proxySelection) throw Error("Couldn't find any selection with the proxy");

        const selections = [...globalSelections].filter((selection) => {
          return selection.selections.has(selection);
        });

        if (selections.length === 0) {
          console.warn("No selections made");
          return Promise.resolve();
        }

        return resolveSelections(selections);
      })()
    );
  }

  async function resolveSelections(selections: Selection[] | Set<Selection>) {
    const { reject, resolve, promise } = deferredPromise();
    for (const selection of selections) {
      const proxySelection = ProxyCache.get(selection);
      if (proxySelection) {
        ProxyResolvePromises.set(selection, promise);
      }
    }

    try {
      const { query, variables } = buildQuery(selections);

      const { data, errors } = await queryFetcher(query, variables);

      if (data) {
        mergeCache(data);

        for (const selection of selections) {
          globalSelections.delete(selection);
        }
      }

      if (errors) {
        console.error(errors);
        const err = Error("Errors in resolve");

        if (Error.captureStackTrace) {
          Error.captureStackTrace(err, resolveAllSelections);
        }

        throw err;
      }

      resolve();
    } catch (err) {
      reject(err);
      throw err;
    }
  }

  async function resolveAllSelections() {
    const selections = [...globalSelections];
    const { reject, resolve, promise } = deferredPromise();
    for (const selection of selections) {
      const proxySelection = ProxyCache.get(selection);
      if (proxySelection) {
        ProxyResolvePromises.set(selection, promise);
      }
    }

    try {
      await resolveSelections(selections);

      resolve();
    } catch (err) {
      reject(err);
      throw err;
    }
  }

  // TODO: Refetch behavior
  function createArrayTypeProxy(schemaType: Schema[string], selectionsArg: Selection) {
    const arrayCacheValue = getCacheFromSelection(selectionsArg);
    if (arrayCacheValue === null) return null;

    const proxy =
      ProxyCache.get(selectionsArg) ||
      (() => {
        const createdProxy: ProxyConstructor = new Proxy(
          arrayCacheValue === CacheNotFound ? [ProxySymbol] : arrayCacheValue,
          {
            get(target, key: string, receiver) {
              const index = parseInt(key);

              if (Number.isInteger(index)) {
                if (arrayCacheValue !== CacheNotFound && arrayCacheValue[index] == null) {
                  // if arrayCacheValue[index] is 'null' or 'undefined'
                  return arrayCacheValue[index];
                }

                const selection = new Selection({
                  key: index,
                  prevSelection: selectionsArg,
                });
                return createTypeProxy(schemaType, selection);
              }
              return Reflect.get(target, key, receiver);
            },
          }
        );
        ProxyCache.set(selectionsArg, createdProxy);
        ProxyCacheReverse.set(createdProxy, selectionsArg);
        return createdProxy;
      })();

    return proxy;
  }

  // TODO: Refetch behavior
  function createTypeProxy(schemaType: Schema[string], selectionsArg: Selection) {
    const cacheValue = getCacheFromSelection(selectionsArg);
    if (cacheValue === null) return null;

    const proxy =
      ProxyCache.get(selectionsArg) ||
      (() => {
        const createdProxy = new Proxy(
          Object.fromEntries(
            Object.keys(schemaType).map((key) => {
              return [key, ProxySymbol];
            })
          ) as Record<string, unknown>,
          {
            get(target, key: string, receiver) {
              if (!schemaType.hasOwnProperty(key)) return Reflect.get(target, key, receiver);

              const value = schemaType[key];
              if (value) {
                const { __type, __args } = value;

                const { pureType, isArray } = (() => {
                  let isNullable = true;
                  let nullableItems = true;
                  let isArray = false;
                  let pureType = __type;
                  if (__type.endsWith("!")) {
                    isNullable = false;
                    pureType = __type.slice(0, __type.length - 1);
                  }
                  if (pureType.startsWith("[")) {
                    pureType = pureType.slice(1, pureType.length - 1);
                    isArray = true;
                    if (pureType.endsWith("!")) {
                      nullableItems = false;
                      pureType = pureType.slice(0, pureType.length - 1);
                    }
                  }

                  return {
                    pureType,
                    isNullable,
                    nullableItems,
                    isArray,
                  };
                })();

                const selection = new Selection({
                  key,
                  prevSelection: selectionsArg,
                  isArray,
                });

                const resolve = (): unknown => {
                  if (scalars[pureType]) {
                    const cacheValue = getCacheFromSelection(selection);

                    if (cacheValue === CacheNotFound) {
                      globalSelections.add(selection);

                      return null;
                    }

                    return cacheValue;
                  }

                  const typeValue = schema[pureType];
                  if (typeValue) {
                    if (isArray) {
                      return createArrayTypeProxy(typeValue, selection);
                    }
                    return createTypeProxy(typeValue, selection);
                  }

                  throw Error("97 Not found!");
                };

                if (__args) {
                  return (args: typeof __args) => {
                    selection.args = args;
                    selection.argTypes = __args;

                    return resolve();
                  };
                }

                return resolve();
              }
              console.error("Not found", key);

              throw Error(`83. Not found`);
            },
          }
        );
        ProxyCache.set(selectionsArg, createdProxy);
        ProxyCacheReverse.set(createdProxy, selectionsArg);
        return createdProxy;
      })();

    return proxy;
  }

  function createSchemaProxy() {
    return new Proxy(
      Object.fromEntries(
        Object.keys(schema).map((key) => {
          return [key, ProxySymbol];
        })
      ) as Record<string, unknown>,
      {
        get(_target, key: string, _receiver) {
          const value = schema[key];

          if (value) {
            const selection = new Selection({
              key,
            });

            return createTypeProxy(value, selection);
          }
          throw Error("104. Not found");
        },
      }
    ) as GeneratedSchema;
  }

  return {
    client,
    globalSelections,
    resolveAllSelections,
    resolve,
  };
}
