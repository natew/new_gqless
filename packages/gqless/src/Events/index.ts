import { ExecutionResult } from 'graphql';

import { CacheType } from '../Cache';
import { gqlessError } from '../Error';
import { Selection } from '../Selection';

export interface FetchEventData {
  executionResult?: Pick<ExecutionResult, 'data' | 'extensions'>;
  error?: gqlessError;
  query: string;
  variables: Record<string, unknown> | undefined;
  cacheSnapshot: CacheType;
  selections: Selection[];
  type: 'query' | 'mutation' | 'subscription';
}

export interface CacheChangeEventData {
  selection: Selection;
  data: unknown;
}

interface OnFetchEventFn {
  (data: Promise<FetchEventData>): void;
}

interface OnCacheChangeEventFn {
  (data: CacheChangeEventData): void;
}

export class EventHandler {
  public hasFetchSubscribers = false;
  private onFetchListeners = new Set<OnFetchEventFn>();

  private onCacheChangeListeners = new Set<OnCacheChangeEventFn>();

  public sendCacheChange(data: CacheChangeEventData) {
    this.onCacheChangeListeners.forEach((listener) => {
      listener(data);
    });
  }

  public sendFetchPromise(data: Promise<FetchEventData>) {
    this.onFetchListeners.forEach((listener) => {
      listener(data);
    });
  }

  public onCacheChangeSubscribe(fn: OnCacheChangeEventFn) {
    const self = this;

    self.onCacheChangeListeners.add(fn);

    return function unsubscribe() {
      self.onCacheChangeListeners.delete(fn);
    };
  }

  public onFetchSubscribe(fn: OnFetchEventFn) {
    const self = this;

    self.hasFetchSubscribers = Boolean(self.onFetchListeners.add(fn).size);

    return function unsubscribe() {
      self.onFetchListeners.delete(fn);
      self.hasFetchSubscribers = Boolean(self.onFetchListeners.size);
    };
  }
}
