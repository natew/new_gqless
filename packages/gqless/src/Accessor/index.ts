import fromPairs from 'lodash/fromPairs';
import { CacheNotFound } from '../Cache';
import { InnerClientState } from '../Client/client';
import { DeepPartial, parseSchemaType, Schema } from '../Schema';
import { Selection, SelectionType } from '../Selection';
import { isInteger } from '../Utils';

export function AccessorCreators<
  GeneratedSchema extends {
    query: {};
    mutation: {};
    subscription: {};
  } = never
>(innerState: InnerClientState) {
  const {
    accessorCache,
    selectionManager,
    interceptorManager,
    scalarsEnumsHash,
    schema,
    eventHandler,
  } = innerState;

  const ProxySymbol = Symbol('gqless-proxy');

  const ResolveInfoSymbol = Symbol();

  interface ResolveInfo {
    key: string | number;
    prevSelection: Selection;
    argTypes: Record<string, string> | undefined;
  }

  const proxySymbolArray = [ProxySymbol];

  function extractDataFromProxy(value: unknown): unknown {
    if (accessorCache.isProxy(value)) {
      const accessorSelection = accessorCache.getProxySelection(value);

      // An edge case hard to reproduce
      /* istanbul ignore if */
      if (!accessorSelection) {
        return undefined;
      }

      const selectionCache = innerState.clientCache.getCacheFromSelection(
        accessorSelection
      );

      if (selectionCache === CacheNotFound) {
        return undefined;
      }
      return selectionCache;
    } else {
      return value;
    }
  }

  function setCache<A extends object>(
    accessor: A,
    data: DeepPartial<A> | null | undefined
  ): void;
  function setCache<B extends (args?: any) => unknown>(
    accessor: B,
    args: Parameters<B>['0'],
    data: DeepPartial<ReturnType<B>> | null | undefined
  ): void;
  function setCache(
    accessor: unknown,
    dataOrArgs: Record<string, unknown> | unknown,
    possibleData?: unknown
  ) {
    if (typeof accessor === 'function') {
      if (dataOrArgs !== undefined && typeof dataOrArgs !== 'object') {
        const err = Error('Invalid arguments of type: ' + typeof dataOrArgs);
        /* istanbul ignore else */
        if (Error.captureStackTrace!) {
          Error.captureStackTrace(err, setCache);
        }
        throw err;
      }

      const resolveInfo = (accessor as Function & {
        [ResolveInfoSymbol]?: ResolveInfo;
      })[ResolveInfoSymbol];

      if (!resolveInfo) {
        const err = Error('Invalid gqless function');

        /* istanbul ignore else */
        if (Error.captureStackTrace!) {
          Error.captureStackTrace(err, setCache);
        }

        throw err;
      }

      const selection = selectionManager.getSelection({
        ...resolveInfo,
        args: dataOrArgs as Record<string, unknown>,
      });

      const data = extractDataFromProxy(possibleData);

      innerState.clientCache.setCacheFromSelection(selection, data);
      eventHandler.sendCacheChange({
        data,
        selection,
      });
    } else if (accessorCache.isProxy(accessor)) {
      const selection = accessorCache.getProxySelection(accessor);

      // An edge case hard to reproduce
      /* istanbul ignore if */
      if (!selection) {
        const err = Error('Invalid proxy selection');

        if (Error.captureStackTrace!) {
          Error.captureStackTrace(err, setCache);
        }

        throw err;
      }

      const data = extractDataFromProxy(dataOrArgs);

      innerState.clientCache.setCacheFromSelection(selection, data);
      eventHandler.sendCacheChange({
        data,
        selection,
      });
    } else {
      const err = Error('Invalid gqless proxy');

      /* istanbul ignore else */
      if (Error.captureStackTrace!) {
        Error.captureStackTrace(err, setCache);
      }

      throw err;
    }
  }

  function createArrayAccessor(
    schemaType: Schema[string],
    selectionArg: Selection
  ) {
    const arrayCacheValue = innerState.clientCache.getCacheFromSelection(
      selectionArg
    );
    if (innerState.allowCache && arrayCacheValue === null) return null;

    const proxyValue: unknown[] =
      arrayCacheValue === CacheNotFound || !Array.isArray(arrayCacheValue)
        ? proxySymbolArray
        : arrayCacheValue;

    const accessor = accessorCache.getArrayAccessor(
      selectionArg,
      proxyValue,
      () => {
        return new Proxy(proxyValue, {
          set(_target, key: string, value: unknown) {
            let index: number | undefined;

            try {
              index = parseInt(key);
            } catch (err) {}

            if (isInteger(index)) {
              const selection = selectionManager.getSelection({
                key: index,
                prevSelection: selectionArg,
              });

              const data = extractDataFromProxy(value);

              innerState.clientCache.setCacheFromSelection(selection, data);

              eventHandler.sendCacheChange({
                selection,
                data,
              });

              return true;
            }

            throw TypeError('Invalid array assignation');
          },
          get(target, key: string, receiver) {
            let index: number | undefined;

            try {
              index = parseInt(key);
            } catch (err) {}

            if (isInteger(index)) {
              if (
                innerState.allowCache &&
                arrayCacheValue !== CacheNotFound &&
                arrayCacheValue[index] == null
              ) {
                /**
                 * If cache is enabled and arrayCacheValue[index] is 'null' or 'undefined', return it
                 */
                return arrayCacheValue[index];
              }

              const selection = selectionManager.getSelection({
                key: index,
                prevSelection: selectionArg,
              });
              const childAccessor = createAccessor(schemaType, selection);

              accessorCache.addAccessorChild(accessor, childAccessor);

              return childAccessor;
            }

            return Reflect.get(target, key, receiver);
          },
        });
      }
    );

    return accessor;
  }

  function createAccessor(schemaType: Schema[string], selectionArg: Selection) {
    const cacheValue = innerState.clientCache.getCacheFromSelection(
      selectionArg
    );
    if (innerState.allowCache && cacheValue === null) return null;

    const accessor = accessorCache.getAccessor(selectionArg, () => {
      return new Proxy(
        fromPairs(
          Object.keys(schemaType).map((key) => {
            return [key, ProxySymbol];
          })
        ) as Record<string, unknown>,
        {
          set(_target, key: string, value: unknown) {
            if (!schemaType.hasOwnProperty(key)) {
              throw TypeError('Invalid proxy assignation');
            }

            const targetSelection = selectionManager.getSelection({
              key,
              prevSelection: selectionArg,
            });

            const data = extractDataFromProxy(value);

            innerState.clientCache.setCacheFromSelection(targetSelection, data);
            eventHandler.sendCacheChange({
              data,
              selection: targetSelection,
            });

            return true;
          },
          get(target, key: string, receiver) {
            if (!schemaType.hasOwnProperty(key))
              return Reflect.get(target, key, receiver);

            const { __type, __args } = schemaType[key];
            const { pureType, isArray } = parseSchemaType(__type);

            const resolve = (args?: {
              argValues: Record<string, unknown>;
              argTypes: Record<string, string>;
            }): unknown => {
              const selection = selectionManager.getSelection({
                key,
                prevSelection: selectionArg,
                args: args != null ? args.argValues : undefined,
                argTypes: args != null ? args.argTypes : undefined,
              });

              if (scalarsEnumsHash[pureType]) {
                const cacheValue = innerState.clientCache.getCacheFromSelection(
                  selection
                );

                accessorCache.addSelectionToAccessorHistory(
                  accessor,
                  selection
                );

                if (cacheValue === CacheNotFound) {
                  // If cache was not found, add the selections to the queue
                  interceptorManager.addSelection(selection);

                  innerState.foundValidCache = false;
                  return null;
                }

                if (!innerState.allowCache) {
                  // Or if you are making the network fetch always
                  interceptorManager.addSelection(selection);
                } else {
                  // For the subscribers of data changes
                  interceptorManager.addSelectionCache(selection);
                }

                return cacheValue;
              }

              const typeValue = schema[pureType];
              if (typeValue) {
                const childAccessor = isArray
                  ? createArrayAccessor(typeValue, selection)
                  : createAccessor(typeValue, selection);

                accessorCache.addAccessorChild(accessor, childAccessor);

                return childAccessor;
              }

              throw Error('GraphQL Type not found!');
            };

            if (__args) {
              const resolveInfo: ResolveInfo = {
                key,
                prevSelection: selectionArg,
                argTypes: __args,
              };

              return Object.assign(
                function ProxyFn(argValues: Record<string, unknown> = {}) {
                  return resolve({
                    argValues,
                    argTypes: __args,
                  });
                },
                {
                  [ResolveInfoSymbol]: resolveInfo,
                }
              );
            }

            return resolve();
          },
        }
      );
    });
    return accessor;
  }

  function createSchemaAccesor() {
    return (new Proxy(
      {
        query: ProxySymbol,
        mutation: ProxySymbol,
        subscription: ProxySymbol,
      },
      {
        get(target, key: string, receiver) {
          const value = schema[key];

          if (value) {
            let type: SelectionType;
            switch (key) {
              case 'subscription': {
                type = SelectionType.Subscription;
                break;
              }
              case 'mutation': {
                type = SelectionType.Mutation;
                break;
              }
              default: {
                type = SelectionType.Query;
              }
            }
            const selection = selectionManager.getSelection({
              key,
              type,
            });

            return createAccessor(value, selection);
          }

          return Reflect.get(target, key, receiver);
        },
      }
    ) as unknown) as GeneratedSchema;
  }

  return {
    createAccessor,
    createArrayAccessor,
    createSchemaAccesor,

    setCache,
  };
}
