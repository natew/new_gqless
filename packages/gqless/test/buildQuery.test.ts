import {
  parse,
  stripIgnoredCharacters as officialStripIgnoredCharacters,
} from 'graphql';

import { buildQuery } from '../src/QueryBuilder';
import { Selection, SelectionType } from '../src/Selection';

describe('basic', () => {
  test('query', () => {
    const baseSelection = new Selection({
      key: 'query',
      type: SelectionType.Query,
    });

    const selectionA = new Selection({
      key: 'a',
      prevSelection: baseSelection,
    });

    const selectionB = new Selection({
      key: 'b',
      prevSelection: baseSelection,
    });

    const { query, variables } = buildQuery([selectionA, selectionB], {
      type: 'query',
    });

    expect(query.trim().startsWith('query{')).toBeTruthy();

    expect(query).toMatchSnapshot('basic query');

    expect(variables).toBe(undefined);

    expect(() => {
      parse(query);
    }).not.toThrow();

    expect(officialStripIgnoredCharacters(query)).toBe(query);
  });

  test('deep query with unions', () => {
    const baseSelection = new Selection({
      key: 'query',
      type: SelectionType.Query,
    });

    const selectionA = new Selection({
      key: 'a',
      prevSelection: baseSelection,
    });

    const selectionB = new Selection({
      key: 'b',
      prevSelection: selectionA,
    });

    const selectionC = new Selection({
      key: 'c',
      prevSelection: selectionB,
    });

    const selectionD = new Selection({
      key: 'd',
      prevSelection: selectionC,
    });

    const selectionE1 = new Selection({
      key: 'a',
      prevSelection: selectionD,
      unions: ['val1', 'val2'],
    });

    const selectionE2 = new Selection({
      key: 'b',
      prevSelection: selectionD,
      unions: ['val1'],
    });

    const selectionF = new Selection({
      key: 'f',
      prevSelection: selectionE1,
    });

    const { query, variables } = buildQuery([selectionE2, selectionF], {
      type: 'query',
    });

    expect(query.trim().startsWith('query{')).toBeTruthy();

    expect(query).toMatchSnapshot('basic query');

    expect(variables).toBe(undefined);

    expect(() => {
      parse(query);
    }).not.toThrow();

    expect(officialStripIgnoredCharacters(query)).toBe(query);
  });

  test('query args', () => {
    const baseSelection = new Selection({
      key: 'query',
      type: SelectionType.Query,
    });

    const selectionA = new Selection({
      key: 'a',
      prevSelection: baseSelection,
      args: {
        a: 1,
        b: 1,
      },
      argTypes: {
        a: 'Int!',
        b: 'String!',
      },
      alias: 'gqlessAlias_1',
    });

    const selectionB = new Selection({
      key: 'a_b',
      prevSelection: selectionA,
    });

    const selectionC = new Selection({
      key: 'a_c',
      prevSelection: selectionA,
    });

    const selectionD = new Selection({
      key: 'd',
      prevSelection: baseSelection,
    });

    const { query, variables } = buildQuery(
      [selectionB, selectionC, selectionD],
      {
        type: 'query',
      }
    );

    expect(query.trim().startsWith('query(')).toBeTruthy();

    expect(query).toMatchSnapshot('basic query args');

    expect(() => {
      parse(query);
    }).not.toThrow();

    expect(variables).toEqual({ a1: 1, b2: 1 });

    expect(officialStripIgnoredCharacters(query)).toBe(query);
  });

  test('mutation args', () => {
    const baseSelection = new Selection({
      key: 'mutation',
      type: SelectionType.Mutation,
    });

    const selectionA = new Selection({
      key: 'a',
      prevSelection: baseSelection,
      args: {
        a: 1,
        b: 1,
      },
      argTypes: {
        a: 'Int!',
        b: 'String!',
      },
      alias: 'gqlessAlias_1',
    });

    const { query, variables } = buildQuery([selectionA], {
      type: 'mutation',
    });

    expect(query.trim().startsWith('mutation(')).toBeTruthy();

    expect(query).toMatchSnapshot('basic mutation with args');

    expect(() => {
      parse(query);
    }).not.toThrow();

    expect(variables).toEqual({ a1: 1, b2: 1 });

    expect(officialStripIgnoredCharacters(query)).toBe(query);
  });
});
