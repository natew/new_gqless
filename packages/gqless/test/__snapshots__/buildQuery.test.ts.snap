// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic deep query with unions: basic query 1`] = `"query{a{b{c{d{...on val1{b a{f}}...on val2{a{f}}}}}}}"`;

exports[`basic mutation args: basic mutation with args 1`] = `"mutation($a1:Int!$b2:String!){gqlessAlias_1:a(a:$a1 b:$b2)}"`;

exports[`basic query args: basic query args 1`] = `"query($a1:Int!$b2:String!){gqlessAlias_1:a(a:$a1 b:$b2){a_b a_c}d}"`;

exports[`basic query: basic query 1`] = `"query{a b}"`;
